### ISSUE TITLE
Ruleset validation not working when root model is a collection in ASP.NET Core

### ISSUE THREAD

**Issue by iamlazydev** [FluentValidation/FluentValidation#1426]
**Which version of FluentValidation are you using?**
8.6.2

**Which version of ASP.NET are you using?**
ASP.NET Core 3.0/3.1

**Describe the issue that you're having**

**Model and Validator**
```
public class Person
{
        public int Id { get; set; }

        public string Name { get; set; }

        public int Age { get; set; }
}

public class PersonValidator : AbstractValidator<Person>
 {
        public PersonValidator()
        {
            RuleSet("update", () =>
            {
                RuleFor(p => p.Id).NotEmpty();
            });

            RuleFor(p => p.Name).NotEmpty();

            RuleFor(p => p.Age).GreaterThan(15);
        }
 }
```
**Contorller**
```
[Route("api/[controller]")]
[ApiController]
public class PersonController : ControllerBase
{
        [HttpPost("addperson")]
        public IActionResult AddPerson([CustomizeValidator(RuleSet = "default,update")] Person person)
        {
            return Ok(person);
        }

        [HttpPost("addpeople")]
        public IActionResult AddPeople([CustomizeValidator(RuleSet = "default,update")] List<Person> people)
        {
            return Ok(people);
        }
 }
```
I am using ASP.NET Core model validation.
Ruleset validation is working for **AddPerson** action with single object model, but not working for **AddPeople** when model because collection.
Tried [Similar issue](https://github.com/FluentValidation/FluentValidation/issues/952) by defining a collection validator.
```
 public class PersonListValidator : AbstractValidator<List<Person>>
 {
        public PersonListValidator()
        {
            RuleForEach(x => x).SetValidator(new PersonValidator());
        }
 }
```
Though you have mentioned this is not needed when using ASP.NET Core model validation, Ruleset validations worked after adding a collection validator, but other properties validations are executed twice.

FluentValidation configuration and validator registrations are as below

```
 services.AddControllers()
       .AddFluentValidation(fv =>
        {
            fv.ImplicitlyValidateChildProperties = true;
        });

services.AddScoped<IValidator<Person>, PersonValidator>();
services.AddScoped<IValidator<List<Person>>, PersonListValidator>();
```
Am i doing anything wrong?, please help to make ruleset validations work on collections.

PS: Collection validators are only working when controller actions take a List<T>, not working on IEnumerable<T>

**Comment by JeremySkinner** [comment]
Hi, there are a few things going on here that will be causing you issues.


Firstly, you should *not* use both `SetCollectionValidator` and set `ImplicitlyValidateChildProperties ` to true. You should pick one approach or the other, or you will end up with duplicate validator executions as you've already experienced. 

When you set `ImplicitlyValidateChildProperties` to true, ASP.NET will recursively look at all the properties on the incoming model object and attempt to locate and instantiate validators for them automatically. However, if you leave it false, then child properties will only be validated if you use SetValidator/SetCollectionValidator. If you use both, then ASP.NET will be instantiating validators for every property, then each validator will also run its child validators with SetValidator/SetCollectionValidator, so you end up with duplicates.

**You must select one strategy and be consistent in this within your application. Do not use both approaches**

In your case, I would suggest that you SetCollectionValidator and *don't* use ImplicitlyValidateChildProperties because you're using the `CustomizeValidator` attribute with rulesets. This will *only work if you're using a root-level model validator* (ie the PersonListValidator). If you set ImplicitlyValidateChildProperties to true and don't define a root-level validator for List<Person> then although ASP.NET will still run its recursive validation, there isn't actually a root validator being run (ie for the whole collection), so the CustomizeValidator attribute doesn't actually have anything to customize, as no root validator actually exists (only child validators are run). 

** If you need to use CustomizeValidator then you must use a root validator, and turn off ImplicitlyValidateChildProperties **

> Collection validators are only working when controller actions take a List, not working on IEnumerable

This is expected. Your validator registrations need to match the action method type parameter. If your action method parameter is a `List<T>`, then your container registration must also be for `IValidator<List<T>>`, if your action method parameter is an `IEnumerable<T>` then your container registration should be for `IValidator<IEnumerable<T>>`. These always need to match up, otherwise ASP.NET will not be able to match up the types. [I have written about this in more detail here](https://github.com/FluentValidation/FluentValidation/issues/1344#issuecomment-595982096).

So in summary, you'll need to do the following:
- Set `ImplicitlyValidateChildProperties` to false (or just remove the line completely) and always use explicit child validators 
- Make sure that your action method parameter collection type always matches the type in your validator declaration, and that these are concrete types.

Hope that makes sense. Let me know if you need me to clarify anything.

**Comment by iamlazydev** [comment]
Thank you for the quick and detailed explanation. Now it works perfectly with your suggested changes.
Thank you once again for the reply and the awesome **FluentValidation**.
